File: src/main/java/com/salton123/base/BaseSupportSwipeBackFragment.java
package com.salton123.base;

import android.app.Activity;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.IdRes;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.orhanobut.logger.Logger;

import java.io.Serializable;

import io.reactivex.disposables.CompositeDisposable;
import me.yokeyword.fragmentation_swipeback.SwipeBackFragment;

/**
 * User: 巫金生(newSalton@163.com)
 * Date: 2017/6/11 10:50
 * Description: 必须要在onViewCreated中return attachToSwipeBack(mContentView);
 * Updated:
 */
public abstract class BaseSupportSwipeBackFragment extends SwipeBackFragment {

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        setParallaxOffset(0.5f);
    }

    public static <T extends Fragment> T newInstance(Class<T> clz) {
        Bundle bundle = new Bundle();
        T fragment = null;
        try {
            fragment = clz.newInstance();
        } catch (Exception e) {
            e.printStackTrace();
        }
        fragment.setArguments(bundle);
        return fragment;
    }

    public static <T extends Fragment> T newInstance(Class<T> clz, Serializable value) {
        Bundle bundle = new Bundle();
        bundle.putSerializable(ARG_ITEM, value);
        T fragment = null;
        try {
            fragment = clz.newInstance();
        } catch (Exception e) {
            e.printStackTrace();
        }
        fragment.setArguments(bundle);
        return fragment;
    }
    public static <T extends Fragment> T newInstance(Class<T> clz, Parcelable value) {
        Bundle bundle = new Bundle();
        bundle.putParcelable(ARG_ITEM, value);
        T fragment = null;
        try {
            fragment = clz.newInstance();
        } catch (Exception e) {
            e.printStackTrace();
        }
        fragment.setArguments(bundle);
        return fragment;
    }

    protected static final String ARG_ITEM = "arg_item";

    public View mContentView;


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        InitVariable(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        mContentView = LayoutInflater.from(getActivity()).inflate(GetLayout(), null);
        return attachToSwipeBack(mContentView);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        InitViewAndData();
        InitListener();
    }

    public abstract int GetLayout();

    public abstract void InitVariable(Bundle savedInstanceState);

    public abstract void InitViewAndData();

    public abstract void InitListener();


    /**
     * 查找View
     *
     * @param id   控件的id
     * @param <VT> View类型
     * @return
     */
    protected <VT extends View> VT f(@IdRes int id) {
        return (VT) mContentView.findViewById(id);
    }

    /**
     * 获取Activity
     *
     * @param <VT>
     * @return
     */
    protected <VT extends Activity> VT activity() {
        return (VT) _mActivity;
    }

    public void toast(String p_Msg) {
        Toast.makeText(activity(), p_Msg, Toast.LENGTH_SHORT).show();

    }

    /**
     * @return 布局解析器
     */
    public LayoutInflater inflater() {
        LayoutInflater _LayoutInflater = LayoutInflater.from(activity());
        return _LayoutInflater;
    }

    public void log(String p_Msg) {
        Logger.d(p_Msg);
    }

    protected CompositeDisposable mCompositeSubscription = new CompositeDisposable();

    @Override
    public void onDestroy() {
        super.onDestroy();
        mCompositeSubscription.dispose();
    }

}

File: src/main/java/com/salton123/base/BaseSupportSwipeBackFragment.kt
package com.salton123.base

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Parcelable
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.salton123.manager.lifecycle.IFragmentLifeCycle
import com.salton123.util.LogUtils
import com.salton123.util.ViewUtils
import me.yokeyword.fragmentation_swipeback.SwipeBackFragment
import java.io.Serializable

/**
 * User: 巫金生(newSalton@163.com)
 * Date: 2017/6/11 10:50
 * Description: 必须要在onViewCreated中return attachToSwipeBack(mContentView);
 * Updated:
 */
abstract class BaseSupportSwipeBackFragment : SwipeBackFragment(), IComponentLife {
    override fun onCreate(savedInstanceState: Bundle?) {
        initVariable(savedInstanceState)
        super.onCreate(savedInstanceState)
        setParallaxOffset(0.5f)
        IFragmentLifeCycle.Factory.get().init(fragmentManager)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return getRootView()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initViewAndData()
        initListener()
    }

    override fun onDestroy() {
        IFragmentLifeCycle.Factory.get().unInit()
        super.onDestroy()
    }

    override fun initListener() {

    }

    override fun context(): Context {
        return context!!
    }

    override fun log(msg: String) {
        LogUtils.d(msg)
    }


    override fun longToast(toast: String) {
        Toast.makeText(activity, toast, Toast.LENGTH_LONG).show()
    }

    override fun shortToast(toast: String) {
        Toast.makeText(activity, toast, Toast.LENGTH_SHORT).show()
    }

    override fun <VT : View> f(id: Int): VT {
        return ViewUtils.f(getRootView(), id)
    }

    override fun getRootView(): View {
        return inflater().inflate(getLayout(), null)
    }

    override fun inflater(): LayoutInflater {
        return LayoutInflater.from(context)
    }

    override fun openActivity(clz: Class<*>, bundle: Bundle?) {
        startActivity(Intent(context, clz).apply { bundle?.let { this@apply.putExtras(it) } })
    }

    override fun openActivityForResult(clz: Class<*>, bundle: Bundle?, requestCode: Int) {
        startActivityForResult(Intent(context, clz).apply { bundle?.let { this@apply.putExtras(it) } }, requestCode)
    }

    companion object {
        open val ARG_ITEM = "arg_item"

        open fun <T : Fragment> newInstance(clz: Class<T>): T? {
            val bundle = Bundle()
            var fragment: T? = null
            try {
                fragment = clz.newInstance()
                fragment!!.arguments = bundle
            } catch (e: Throwable) {

            }

            return fragment
        }

        open fun <T : Fragment> newInstance(clz: Class<T>, bundle: Bundle): T {
            clz.newInstance().arguments
            return clz.newInstance().also { it.arguments = bundle }
        }

        open fun <T : Fragment> newInstance(clz: Class<T>, value: Serializable): T {
            return clz.newInstance().also {
                it.arguments = Bundle().also {
                    it.putSerializable(ARG_ITEM, value)
                }
            }
        }

        open fun <T : Fragment> newInstance(clz: Class<T>, value: Parcelable): T {
            return clz.newInstance().also {
                it.arguments = Bundle().also {
                    it.putParcelable(ARG_ITEM, value)
                }
            }
        }
    }
}

